ts/ui/safe/safe.ts

```diff
@@ -0,0 +1,267 @@
+
+/**
+ * The Safe object for sanitizing the internal MathML representation of an expression
+ *
+ * @template N  The HTMLElement node class
+ * @template T  The Text node class
+ * @template D  The Document class
+ */
+export class Safe<N, T, D> {
+
+  /**
+   * The options controlling the handling of the safe extension
+   */
+  public static OPTIONS: OptionList = {
+    allow: {
+      //
+      //  Values can be "all", "safe", or "none"
+      //
+      URLs:    'safe',   // safe are in safeProtocols below
+      classes: 'safe',   // safe start with MJX-
+      cssIDs:  'safe',   // safe start with MJX-
+      styles:  'safe'    // safe are in safeStyles below
+    },
+    //
+    // Largest padding/border/margin, etc. in em's
+    //
+    lengthMax: 3,
+    //
+    // Valid range for scriptsisemultiplier
+    //
+    scriptsizemultiplierRange: [.6, 1],
+    //
+    // Valid range for scriptlevel
+    //
+    scriptlevelRange: [-2, 2],
+    //
+    // Pattern for allowed class names
+    //
+    classPattern: /^mjx-[-a-zA-Z0-9_.]+$/,
+    //
+    // Pattern for allowed ids
+    //
+    idPattern: /^mjx-[-a-zA-Z0-9_.]+$/,
+    //
+    // Pattern for data attributes
+    //
+    dataPattern: /^data-mjx-/,
+    //
+    //  Which URL protocols are allowed
+    //
+    safeProtocols: expandable({
+      http: true,
+      https: true,
+      file: true,
+      javascript: false,
+      data: false
+    }),
+    //
+    //  Which styles are allowed
+    //
+    safeStyles: expandable({
+      color: true,
+      backgroundColor: true,
+      border: true,
+      cursor: true,
+      margin: true,
+      padding: true,
+      textShadow: true,
+      fontFamily: true,
+      fontSize: true,
+      fontStyle: true,
+      fontWeight: true,
+      opacity: true,
+      outline: true
+    }),
+    //
+    //  CSS styles that have Top/Right/Bottom/Left versions
+    //
+    styleParts: expandable({
+      border: true,
+      padding: true,
+      margin: true,
+      outline: true
+    }),
+    //
+    //  CSS styles that are lengths needing max/min testing
+    //    A string value means test that style value;
+    //    An array gives [min,max] in em's
+    //    Otherwise use [-lengthMax,lengthMax] from above
+    //
+    styleLengths: expandable({
+      borderTop: 'borderTopWidth',
+      borderRight: 'borderRightWidth',
+      borderBottom: 'borderBottomWidth',
+      borderLeft: 'borderLeftWidth',
+      paddingTop: true,
+      paddingRight: true,
+      paddingBottom: true,
+      paddingLeft: true,
+      marginTop: true,
+      marginRight: true,
+      marginBottom: true,
+      marginLeft: true,
+      outlineTop: true,
+      outlineRight: true,
+      outlineBottom: true,
+      outlineLeft: true,
+      fontSize: [.707, 1.44]
+    })
+  };
+
+  /**
+   * The attribute-to-filter-method mapping
+   */
+  public filterAttributes: {[name: string]: string} = {
```

Comment:
I think this should be a `Map`. This would avoid all the `hasOwnProperty` tests later.

Reply:
Changed to a Map